### PR TITLE
Honour user's choice of ServerProvider

### PR DESCRIPTION
**Pull request by ignasi35**

### CODE REVIEW

**Comment by mkurz** [comment]
Many commits... Something went wrong here.

**Comment by ignasi35** [comment]
> Many commits... Something went wrong here.

it's past my gremlins-commit-hour. 😅

PR'd to the wrong branch.

File: framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala [hunk 1]
```diff
@@ -70,20 +68,42 @@ import play.core.server.SelfSignedSSLEngineProvider
...
+    ServerProvider.fromConfiguration(getClass.getClassLoader, serverConfig(app).configuration)
 
   protected def serverEndpoints(testServer: TestServer): ServerEndpoints = {
-    val httpEndpoint: Option[ServerEndpoint] = testServer.runningHttpPort.map(_ =>
-      ServerEndpointRecipe.AkkaHttp11Plaintext.createEndpointFromServer(testServer)
-    )
-    val httpsEndpoint: Option[ServerEndpoint] = testServer.runningHttpsPort.map(_ =>
-      ServerEndpointRecipe.AkkaHttp20Encrypted.createEndpointFromServer(testServer)
-    )
+    val useAkkaHttp = testServer.serverProvider.get.isInstanceOf[AkkaHttpServerProvider]
```
Comment (dwijnand) [inline framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala]: Ouch.  No better way than `isInstanceOf`?

File: framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala [hunk 2]
```diff
@@ -70,20 +68,44 @@ import play.core.server.SelfSignedSSLEngineProvider
...
+      val recipe = if (useAkkaHttp) {
+        if (useHttp2) {
+          ServerEndpointRecipe.AkkaHttp20Encrypted
+        } else {
+          ServerEndpointRecipe.AkkaHttp11Encrypted
+        }
+      } else {
+        ServerEndpointRecipe.Netty11Encrypted
+      }
+      recipe.createEndpointFromServer(testServer)
```
Comment (dwijnand) [inline framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala]: How painful...

**Comment by ignasi35** [comment]
> Weird, there's no build for this PR in https://travis-ci.org/playframework/playframework/pull_requests.

When created the PR from `hub` the target branch was autoselected to `playframework:master`. I later edited the target branch (thanks @mkurz). I wonder if that messed up travis. I'll close/reopen to see if that triggers a new build.

File: framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala (hunk 1 above)
Comment (ignasi35) [reply framework/src/play-test/src/main/scala/play/api/test/TestServerFactory.scala]: Yeah, very not nice. We will have to fix in a new PR (related to https://github.com/akka/akka-grpc/issues/461)

**Comment by mkurz** [comment]
This needs to be ported to `master`.

**Comment by dwijnand** [comment]
8babe412a769da3c90108c51cd6c33260afef92b

**Pull request merged**